### DISCUSSION TITLE
STM32F417, Host MSC tuh_msc_read10 problem

### DISCUSSION THREAD

**Discussion by linnet1967** [hathach/tinyusb#1032] [Q&A]
Hello,

We have an in-house platform with our own HAL and Synoptics USB driver for low level USB and HCD. I am integrating tinyusb and so far has been ok. I have integrated Chan's FatFS on top of tinyusb. At the moment I have not lifted the OS yet so I am running "bare metal".

Inquiry and finding the device seems fine and I have tested with a couple of different devices.

However, when reading from filesystem I am running into some issues in tuh_msc_read10. The callback does not executed which in turn means that data from device is not read before notification to stack.

Anyone with similar problem?

Best Regards
/s

**Comment by cjhilder** [comment]
Yes, there are other people with similar problems. Trouble is they are also at a loss as to how to solve the problem. I'm hitting the same issue on RP2040, although the cb is being called most of the time, it is definitely not being called about 1 in 7 calls, seems to be non-determinate and random. I'm trying to stitch together a working version from various forks - the host stack is heavily under development and getting a working set of files is a bit of a challenge right now.

**Comment by linnet1967** [comment]
Thanks. I was able to solve it temporarily but later I had to switch onto other parts of our system. I will get back to USB in a couple of months. I remember that there where many problems with the callbacks not called properly and in order. Could be a design flaw or that the build system needs some tweaking.

**Reply by cjhilder** [reply]
I believe I have resolved the callback issues. I'm fighting other weirdness in the USB sub-system, but integration of tinyusb and FatFs is working.

The solution can be found in my tinyusb fork: https://github.com/cjhilder/tinyusb/tree/master/examples/host/msc_with_fatfs

For discussion of the other issues I am facing with the example code see: https://forums.raspberrypi.com/viewtopic.php?t=331298
